### ISSUE TITLE
sql: allow COLLATED column in index definition

### ISSUE THREAD

**Issue by dougmrqs** [cockroachdb/cockroach#63868]
**Is your feature request related to a problem? Please describe.**

Can't pass in COLLATE with a column when creating an INDEX there.

Implementing an adapter for Sequelize, [this test](https://github.com/sequelize/sequelize/blob/main/test/integration/model.test.js#L341) breaks. It errors: 

```
root@:26257/defaultdb> CREATE UNIQUE INDEX "user_and_email_index" ON "UserWithUniqueUsernames" ("user_id", "email" COLLATE "en_US" DESC);
invalid syntax: statement ignored: at or near "collate": syntax error
SQLSTATE: 42601
DETAIL: source SQL:
CREATE UNIQUE INDEX "user_and_email_index" ON "UserWithUniqueUsernames" ("user_id", "email" COLLATE "en_US" DESC)
```

**Describe the solution you'd like**

--

**Describe alternatives you've considered**

Quoting:
> I guess a workaround would be to define the column with COLLATE and then add an INDEX on the column instead of adding COLLATE in the INDEX.
@RichardJCai

**Comment by mgartner** [comment]
This looks like an expression-based index, something we are planning on adding in 21.2.

The 21.1 work-around would be to create an index on a virtual column. Unfortunately, the optimizer does not realize it can use such an index to satisfy queries.

```
demo@127.0.0.1:26257/defaultdb> SHOW CREATE TABLE t;
  table_name |                       create_statement
-------------+----------------------------------------------------------------
  t          | CREATE TABLE public.t (
             |     s STRING NULL,
             |     c STRING COLLATE en_US NULL AS (s COLLATE en_US) VIRTUAL,
             |     rowid INT8 NOT VISIBLE NOT NULL DEFAULT unique_rowid(),
             |     CONSTRAINT "primary" PRIMARY KEY (rowid ASC),
             |     INDEX t_c_idx (c ASC),
             |     FAMILY "primary" (s, rowid)
             | )
(1 row)

Time: 7ms total (execution 7ms / network 0ms)

demo@127.0.0.1:26257/defaultdb> EXPLAIN SELECT * FROM t WHERE s COLLATE "en_US" = 'foo' COLLATE "en_US";
                         info
-------------------------------------------------------
  distribution: full
  vectorized: true

  • render
  │
  └── • filter
      │ filter: s COLLATE en_US = 'foo' COLLATE en_US
      │
      └── • scan
            missing stats
            table: t@primary
            spans: FULL SCAN
(12 rows)
```

I've created https://github.com/cockroachdb/cockroach/issues/65343 to track this issue.

**Comment by mgartner** [comment]
The index on a virtual column works fine as long as you use an unquoted locale in the `COLLATE` expression in the query:

```
defaultdb> CREATE TABLE t (s STRING, c STRING COLLATE en_US AS (s COLLATE en_US) VIRTUAL, INDEX (c));
CREATE TABLE

defaultdb> EXPLAIN SELECT * FROM t WHERE s COLLATE en_US = 'foo' COLLATE en_US;
                               info
------------------------------------------------------------------
  distribution: local
  vectorized: true

  • render
  │
  └── • index join
      │ table: t@primary
      │
      └── • scan
            missing stats
            table: t@t_c_idx
            spans: [/'foo' COLLATE en_US - /'foo' COLLATE en_US]
(12 rows)
```

So, the work-around mentioned should work for now until expression-based indexes are supported (targetting 21.2).

**Comment by mgartner** [comment]
Expression indexes have landed on `master` and they will be released in version 21.2. Here's an example:

```
CREATE TABLE t (s STRING, INDEX ((s COLLATE en_US)));

EXPLAIN SELECT * FROM t WHERE s COLLATE en_US = 'foo' COLLATE en_US;
                             info
--------------------------------------------------------------
  distribution: local
  vectorized: true

  • index join
  │ table: t@primary
  │
  └── • scan
        missing stats
        table: t@t_expr_idx
        spans: [/'foo' COLLATE en_US - /'foo' COLLATE en_US]
```

I'll close the issue.
